### DISCUSSION TITLE
Usage of `futures::executor::block_on`

### DISCUSSION THREAD

**Discussion by paologallinaharbur** [kube-rs/kube#1354] [Q&A]
We are trying to call async method hiding calls to k8s from no-async  methods:

Es:
```rust
    fn get(&self, agent_id: &str) -> Result<Option<DataStored>, StorerError> {
        futures::executor::block_on(self.async_get(agent_id))
    }

    async fn async_get(&self, agent_id: &str) -> Result<Option<DataStored>, StorerError> {
       [...]
        let data_res = self
            .k8s_executor
            .get_configmap_key()
            .await?;
```

and we are in a Tokio runtime marked with `#[tokio::main]`

I initially had some issues, but now it seems to be working. However, while researching I found the following comment in an [old issue](https://github.com/kube-rs/kube/issues/849#issuecomment-1068463927) from @nightkr

It states:
```
You can use tokio::Runtime::block_on, but not futures::block_on 
(at least not without doing a bunch of extra setups to make sure that you still have a Tokio runtime in scope).
```

Are there any issues I am not seeing?  are they compatible?

**Comment by nightkr** [comment]
So there are several issues going on here, some of which hide the others...

1. Tokio features largely need to have a Tokio runtime available. `futures::block_on` does _not_ provide this by itself, however in your case it would be inherited from the `tokio::main`. `tokio::Runtime::block_on` _does_ provide this. You can also use `tokio::Runtime::enter` yourself to install one if you want to use a different executor.
2. Running blocking IO code in an async context blocks the entire thread, which prevents other futures from progressing (either ones that are explicitly joined using something like `futures::future::{select,join}`, ones running in a single-threaded runtime, or ones that are `tokio::spawn`ed but happen to be scheduled for the same thread). `tokio::Runtime::block_on` will detect this case and panic, while `futures::block_on` will apparently leave you to glue your foot back on yourself.
3. Frankly, trying to hide futures like this makes your code much less clear. `block_on` should _only_ be used when trying to run a piece of async code in an otherwise async-unaware application.

All in all, this smells like a case of the XY problem. What is the actual goal that you are trying to accomplish here?
